### PR TITLE
Role icon for admins/moderators and Streak icon according to user activity

### PR DESCRIPTION
**Pull request by ananyo2012**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] tests pass -- `rake test`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request are descriptively named
* [x] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer
* [x] `development.sqlite.example` has been updated if any database migrations were added

Isuue #215 , #319 
Thanks!

### CODE REVIEW

**Comment by ananyo2012** [comment]
@jywarren Please see 9dda564 for #215. See 0ba38ba for #319. 
**Explanation for `user.streak` Method**
I have checked the number for X consecutive days for notes,wiki edits, comments. Then stored the no. of notes, wiki edits and comments in an array. The streak is the maximum no. of days within these three basically the X consecutive days the user has been active. If the streak is greater than a week a fire icon is shown beside the bio pic with the tooltip explaining it. Do we need to check whether there are Y no. of posts within these X days and then only show the icon as you mentioned previously?

**Comment by jywarren** [comment]
Hi, Ananyo - this looks cool -- I think we should try it out, but it needs a few changes first:

* [ ] instead of using new image assets, please use [FontAwesome icons](http://fontawesome.io/icon/fire/), which are already included: `<a href="/wiki/streaks"><i class="fa fa-fire"></i></a>`
* [ ] let's put the icon `class="pull-right"` on the right edge of the window 
* [ ] let's link to a page explaining what this is: `<a href="/wiki/streaks">icon</a>` 
* [ ] let's use a Bootstrap tooltip to indicate what the icon means: https://getbootstrap.com/javascript/#tooltip "On a streak!"

Thanks Ananyo!

**Comment by ananyo2012** [comment]
@jywarren Made the changes asked. Please review.

**Comment by jywarren** [comment]
This looks good in the code, but I'd like to take a few more days to think through the design; i can add a commit on top before merging, and will take a look on Monday. 

I'd like input from @steviepubliclab and @ebarry on both these questions:

* the icons to represent "admins" "moderators" and "being on a streak"
* the definition of "being on a streak" in plain language so that we can display the meaning to people

Ananyo, can you describe as clearly as possible the conditions that cause a "streak" as you've implemented it?

Thank you, this is an advanced and interesting feature and we really appreciate your contribution!

**Comment by jywarren** [comment]
screenshot 2016-05-21 at 10 37 42 am

@steviepubliclab and @ebarry -- available icons are here: https://fontawesome.io/icons/

**Comment by ananyo2012** [comment]
**Explanation for User Streaks**
So here goes the meaning of streak in simple language.Streak refers to the number of consecutive days a user has been active in the community.Activity is measured in terms of notes,wiki edits and comments.If the user has posted even a single note, comment or wiki edits in a particular day he is marked as "active" on that day. If the user is active for more than 7 consecutive days a fire icon is displayed beside his bio pic. So no. of days in streak is the no. of consecutive days the user has been active. The total no. of notes, comments and wiki edits are counted in these days and displayed in the tooltip accordingly. So suppose the user has been "active" for 30 consecutive days with 50 notes, 100 comments and 30 wiki edits in these 30 days then the tootlip shows:
`You are on fire! 50 notes, 30 wiki edits, 100 wiki edits in last 30 days`

**Comment by ananyo2012** [comment]
**Feedback**
Please give your feedback on what the meaning of streak should be. One thing I would like to mention is I have not used any hard margin for the no of notes/wiki edits/comments posted on each day. So even if a user posts just one comment on a particular day(with no notes or wiki edits) he is marked "active". So if we need to do so we need to change the code accordingly.

**Comment by ebarry** [comment]
Wow this is awesome! Love the tooltip containing all the info. 

Re: "definition of streak" -- Being active for 7 consecutive days seems reasonable -- is there a way to check currently how many users would be on a streak under that definition and length of time? I really have no idea whether that's actually high or low for our community activity. Ooooo -- should we send out an informal poll to our community or maybe just organizers "what is a streak?"

Re: icons. So overall, i think DesignWG should weigh in on this one. Here are some ideas tho:

Streak:
* http://fontawesome.io/icon/fire fire is probably my favorite. Can we add a color? like purple?
* other option http://fontawesome.io/icon/line-chart/ indicates activity "trending up"
* other option http://fontawesome.io/icon/level-up/ could be cute

Moderator: 
* http://fontawesome.io/icon/life-ring the alias for this is "support" which seems appropriate
* http://fontawesome.io/icon/heart-o
* http://fontawesome.io/icon/heart

Admin: not having a lot of ideas on this
* http://fontawesome.io/icon/hand-spock-o
* http://fontawesome.io/icon/shield (not a fan, looks authoritative, and/or medieval)
* also consider the options listed above for "Moderator"



I will patiently await the appropriate time to deploy http://fontawesome.io/icon/hand-lizard-o

**Comment by jywarren** [comment]
OMG "hand-spock-o" is pretty amazing.

Let's also render the hand in a lighter color -- maybe #888.

**Comment by ananyo2012** [comment]
@ebarry Yeah it would be great if we would put this up in the community and ask for their definition of streak. Well I like like the shield icon for an admin and life ring sees awesome for a moderator!

**Comment by jywarren** [comment]
What do you all think of a flag icon for moderators? Loaded, but also recognizable. http://fontawesome.io/icon/flag

Asterisk icon for admins? http://fontawesome.io/icon/asterisk

Or a key? http://fontawesome.io/icon/key

**Comment by ebarry** [comment]
i could be into "key" for admin.

--

+1 336-269-1539
@lizbarry <http://twitter.com/lizbarry>

**Comment by ananyo2012** [comment]
@jywarren I am still in for a shield icon for an admin.

**Comment by jywarren** [comment]
@ebarry - if you want to make a final call, Ananyo can update his PR and I
can review before Friday for merging. Thanks!

**Comment by jywarren** [comment]
I'm going to wait a few more days on this, as neither stevie or liz is working on PL stuff today, and the changes look like they'll be pretty straightforward to rebase over other merges. Thanks for your patience, everyone!

**Comment by ananyo2012** [comment]
@jywarren No problem Jeff I can make the changes after @ebarry gives her final decisions on it.

**Comment by steviepubliclab** [comment]
I'm a little conflicted on separating people out by their activity and roles on PL. I think this is coming from some of the readings that have been passed around about activity in online communities. However, I also see the benefit for people to personally identify and share out about who they are, and what they've been up to. This can also be a positive reinforcer of activity. I think some of they points to keep in mind in building these functions that we don't create
- an intimidating space for newcomers or first time contributors, and 
- an internal hierarchy, or become exclusionary by parsing out groups or contributors. 

I'm conflicted as well because in my mind I also go strait to::  What would people think of adding one for organizers?  

Love to hear other people's thoughts on this. 

@jywarren The flag is used in the "geographies" idea a lot. For example, where there currently are "chapters" on PL. Might change, but I don't want to confuse it.

File: app/views/notes/_comment.html.erb [hunk 1]
```diff
@@ -18,6 +18,17 @@
     </div>
 
     <div class="navbar-text navbar-right pull-right" style="padding-right:10px;">
+      <% if comment.author && comment.author.user && comment.author.user.role %>
+        <% streak = comment.author.user.streak
+           if streak[0] > 7 %>
+          <a href="/wiki/streaks"><i class="fa fa-fire" style="font-size:16px; color:#FF0000;" data-toggle="tooltip" data-placement="top" title="You are on Fire! <%= streak[1][0]%> notes, <%= streak[1][1]%> wiki edits, <%= streak[1][2]%> comments in last <%= streak[0]%> days"></i></a>
+        <% end %>
+        <% if comment.author.user.role == 'admin' %>
```
Comment (jywarren) [inline app/views/notes/_comment.html.erb]: Let's just comment the role icon stuff out for now -- pending further design discussion on #215. Thanks, and sorry it wasn't quite at a design state to be implemented yet! Let's not let it hold up the streaks function.

File: test/functional/notes_controller_test.rb [hunk 2]
```diff
@@ -289,4 +289,26 @@ def setup
     assert_tag :tag => 'iframe', attributes: {src: 'http://mapknitter.org/embed/sattelite-imagery'}
   end
 
+  test "should mark admins and moderators with a special icon" do
```
Comment (jywarren) [inline test/functional/notes_controller_test.rb]: This can also be commented out for now. Thanks!

**Comment by jywarren** [comment]
Hi, thanks - I think we're a little stalled on the role icon part of this, and since the functional part of the code is already completed, I'm going to suggest that we comment out the role icon parts (i've left comments at the relevant parts of the code) and move forward with the "streaks". Thanks, Ananyo!

**Comment by ananyo2012** [comment]
@jywarren Ok following up as suggested.

**Comment by ananyo2012** [comment]
@jywarren I have commented out the code for role icons as asked. I also rebased it on top of the current master. I hope this can be merged now.

**Comment by jywarren** [comment]
Making some small style tweaks and merging manually -- thanks, Ananyo, this is great! Closing. 

screenshot 2016-06-03 at 4 43 29 pm

**Comment by Fastie** [comment]
I agree with Stevie that the type of streak defined here will reinforce the distinction between elite users and the 99% of Public Lab folks. The type of streak that might be more suitable could be based on how much community interaction is related to a user's contributions. So, e.g., how many comments  and likes on your research notes, how many research notes posted as a response to your notes, how many times your username was mentioned in a note or comment, how many barnstars were awarded to you, maybe how many comments on notes that you also commented on. A streak could be some minimum number of these things during a week -- consecutive days might be too restrictive. A displayed icon alerting you that you are on a streak is good, and there should also be a displayed record of when you were on a streak in the past.

**Comment by jywarren** [comment]
Hi, I was also thinking the current streak definition may be too difficult.
The goal is to encourage new contributors, so can you suggest a streak
definition that'd help?

**Comment by Fastie** [comment]
To encourage new contributors, just multiply everybody's streak score by a factor related to newness of the user, e.g., the inverse of the number of days since joining or since first posting. To account for people who posted once three years ago but not since, wipe the slate clean if a user has not posted in six months (six weeks?). This metric would need a new name because brand new users would score high with just two or three comments, which might not be much of a "streak."

**Comment by jywarren** [comment]
Oh, I like that, I was actually thinking that perhaps you could only get a
streak if you were new, but maybe a proportional streak would work?

**Comment by ananyo2012** [comment]
There are some really great thoughts on the streak feature. It would be great if everyone were to give their views on this as @ebarry mentioned. Should I post a Question and ask for a response in the community? Then we can come to a conclusion what should be right steps to decide an user on streak.

**Comment by jywarren** [comment]
yes, that's a great idea, ananyo!

**Comment by jywarren** [comment]
Also, ananyo - apologies, but I introduced a bug where now *everyone* is on a streak. I'll fix this and publish soon, but it also highlighted the fact that the tooltip says "You're on fire" -- even if the user with the streak is not you. This may also have been an oversight on my part, I don't remember -- i was traveling all last week, so perhaps moved a little too fast. I'll include this in my fix.

**Comment by ebarry** [comment]
@steviepubliclab and I are working together today and want to suggest that this idea of "proportional streak" could be implemented to be shown privately to the user to encourage them, but not displayed publicly.

**Comment by ananyo2012** [comment]
We can perhaps show the streaks of an user in the profile page. Like github used to show earlier.

**Pull request closed without merging**